osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/infra/messaging/processors/KeysRequestMessageProcessor.java

```diff
@@ -0,0 +1,108 @@
+import org.opensmartgridplatform.adapter.protocol.dlms.application.services.DeviceKeyProcessingService;
+import org.opensmartgridplatform.adapter.protocol.dlms.domain.entities.DlmsDevice;
+import org.opensmartgridplatform.adapter.protocol.dlms.domain.factories.DlmsConnectionManager;
+import org.opensmartgridplatform.adapter.protocol.dlms.exceptions.DeviceKeyProcessAlreadyRunningException;
+import org.opensmartgridplatform.adapter.protocol.dlms.infra.messaging.DeviceRequestMessageProcessor;
+import org.opensmartgridplatform.adapter.protocol.dlms.infra.messaging.DeviceRequestMessageSender;
+import org.opensmartgridplatform.dto.valueobjects.smartmetering.SetKeysRequestDto;
+import org.opensmartgridplatform.shared.exceptionhandling.OsgpException;
+import org.opensmartgridplatform.shared.infra.jms.MessageMetadata;
+import org.opensmartgridplatform.shared.infra.jms.MessageType;
+import org.springframework.beans.factory.annotation.Autowired;
+
+@Slf4j
+public class KeysRequestMessageProcessor extends DeviceRequestMessageProcessor {
+
+  @Autowired private ConfigurationService configurationService;
+
+  @Autowired private DeviceRequestMessageSender deviceRequestMessageSender;
+
+  @Autowired private DeviceKeyProcessingService deviceKeyProcessingService;
+
+  public KeysRequestMessageProcessor(final MessageType messageType) {
+    super(messageType);
+  }
+
+  @Override
+  protected Serializable handleMessage(
+      final DlmsConnectionManager conn,
+      final DlmsDevice device,
+      final Serializable requestObject,
+      final MessageMetadata messageMetadata)
+      throws OsgpException {
+
+    this.assertRequestObjectType(SetKeysRequestDto.class, requestObject);
+
+    final SetKeysRequestDto keySet = (SetKeysRequestDto) requestObject;
+    this.configurationService.replaceKeys(conn, device, keySet, messageMetadata);
+
+    return null;
+  }
+
+  @Override
+  public void processMessage(final ObjectMessage message) throws JMSException {
+    final MessageMetadata messageMetadata = MessageMetadata.fromMessage(message);
+    final Serializable messageObject = message.getObject();
+
+    log.debug(
+        "Checking if not already processing key for device {} before processing {} request message",
+        messageMetadata.getDeviceIdentification(),
+        messageMetadata.getMessageType());
+
+    try {
+      final DlmsDevice device = this.domainHelperService.findDlmsDevice(messageMetadata);
+
+      this.checkIfOtherKeyProcessIsRunning(device.getDeviceIdentification());
+
+      this.markKeyProcessIsRunning(device.getDeviceIdentification());
+
+      super.processMessage(message);
+
+    } catch (final DeviceKeyProcessAlreadyRunningException exception) {
+
+      final Duration deviceKeyProcessingTimeout =
+          this.deviceKeyProcessingService.getDeviceKeyProcessingTimeout();
+      log.debug(
+          "Key(s) already being processed for device {}. {} request message is send back to the MessageSender with delay of {} seconds.",
```

Comment:
Typo "is send back" should be "is sent back", and it is not sent back to the message sender, it is sent back by the message sender to the queue where it was picked up from (as written in the implementation comment a few lines down).

Reply:
log message rewritten